worker/src/router.rs

```diff
@@ -264,67 +271,93 @@ impl<'a, D: 'static> Router<'a, D> {
+                .expect(&format!(
+                    "failed to register {:?} route for {} pattern",
+                    method, pattern
+                ));
         }
     }
 
     /// Handle the request provided to the `Router` and return a `Future`.
     pub async fn run(self, req: Request, env: Env) -> Result<Response> {
-        let (handlers, data) = self.split();
+        let (handlers, data, not_found_handler) = self.split();
+
+        if let Some(handlers) = handlers.get(&req.method()) {
+            if let Ok(Match { value, params }) = &handlers.at(&req.path()) {
+                let mut par: RouteParams = HashMap::new();
+                for (ident, value) in params.iter() {
+                    par.insert(ident.into(), value.into());
+                }
+                let route_info = RouteContext {
+                    data,
+                    env,
+                    params: par,
+                };
+                return match value {
+                    Handler::Sync(func) => (func)(req, route_info),
+                    Handler::Async(func) => (func)(req, route_info).await,
+                };
+            }
+        }
 
-        if let Ok(Match { value, params }) = handlers.at(&req.path()) {
-            let mut par: RouteParams = HashMap::new();
-            for (ident, value) in params.iter() {
-                par.insert(ident.into(), value.into());
+        for method in Method::all() {
+            if method == Method::Head || method == Method::Options || method == Method::Trace {
+                continue;
+            }
+            if let Some(handlers) = handlers.get(&method) {
+                if let Ok(Match { .. }) = handlers.at(&req.path()) {
+                    return Response::error("Method Not Allowed", 405);
+                }
             }
+        }
+
+        if let Some(handler) = not_found_handler {
             let route_info = RouteContext {
                 data,
                 env,
-                params: par,
+                params: HashMap::new(),
+            };
+            return match handler {
+                Handler::Sync(func) => (func)(req, route_info).map(|resp| resp.with_status(404)),
+                Handler::Async(func) => (func)(req, route_info)
+                    .await
+                    .map(|resp| resp.with_status(404)),
             };
-
-            if let Some(handler) = value[req.method() as usize].as_ref() {
-                return match handler {
-                    Handler::Sync(func) => (func)(req, route_info),
-                    Handler::Async(func) => (func)(req, route_info).await,
-                };
-            }
-            return Response::error("Method Not Allowed", 405);
         }
+
         Response::error("Not Found", 404)
     }
 }
 
-type NodeWithHandlers<'a, D> = Node<[Option<Handler<'a, D>>; 9]>;
+type NodeWithHandlers<'a, D> = Node<Handler<'a, D>>;
 
 impl<'a, D: 'static> Router<'a, D> {
```

Comment:
Hmm, why does this have a `'static` bound?

Reply:
The `Handler::Async` enum variant keeps hold of a `fn(Request, RouteContext<D>` and otherwise complains about `D` maybe not living long enough. Feel free to take a closer look and see if there is a way around it!

Reply:
maybe it could just be `'a` though... will test it out

Reply:
nope. still yelled at :) not my top priority to change this up right now, but again feel free to re-work it if you have time and want to think about a different approach

Reply:
Hmm, how are you testing this? I ran `cargo check --all-targets` with this diff and it seemed to compile ok:
```
diff --git a/worker/src/router.rs b/worker/src/router.rs
index c26ba54..67957cd 100644
--- a/worker/src/router.rs
+++ b/worker/src/router.rs
@@ -350,7 +350,7 @@ impl<'a, D: 'static> Router<'a, D> {
 
 type NodeWithHandlers<'a, D> = Node<Handler<'a, D>>;
 
-impl<'a, D: 'static> Router<'a, D> {
+impl<'a, D> Router<'a, D> {
     fn split(
         self,
     ) -> (
```

Reply:
In fact the rest of the API doesn't need to require 'static either:
```diff
diff --git a/worker/src/router.rs b/worker/src/router.rs
index c26ba54..6bbdb82 100644
--- a/worker/src/router.rs
+++ b/worker/src/router.rs
@@ -16,7 +16,7 @@ use crate::{
 
 type HandlerFn<D> = fn(Request, RouteContext<D>) -> Result<Response>;
 type AsyncHandlerFn<'a, D> =
-    Rc<dyn Fn(Request, RouteContext<D>) -> LocalBoxFuture<'a, Result<Response>>>;
+    Rc<dyn 'a + Fn(Request, RouteContext<D>) -> LocalBoxFuture<'a, Result<Response>>>;
 
 /// Represents the URL parameters parsed from the path, e.g. a route with "/user/:id" pattern would
 /// contain a single "id" key.
@@ -89,7 +89,7 @@ impl<D> RouteContext<D> {
     }
 }
 
-impl<'a, D: 'static> Router<'a, D> {
+impl<'a, D: 'a> Router<'a, D> {
     /// Construct a new `Router`, with arbitrary data that will be available to your various routes.
     /// If no data is needed, provide any valid data. The unit type `()` is a good option.
     pub fn new(data: D) -> Self {
@@ -159,7 +159,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// `async/await` syntax in the callback.
     pub fn head_async<T>(mut self, pattern: &str, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -173,7 +173,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// `async/await` syntax in the callback.
     pub fn get_async<T>(mut self, pattern: &str, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -187,7 +187,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// `async/await` syntax in the callback.
     pub fn post_async<T>(mut self, pattern: &str, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -201,7 +201,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// `async/await` syntax in the callback.
     pub fn put_async<T>(mut self, pattern: &str, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -215,7 +215,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// `async/await` syntax in the callback.
     pub fn patch_async<T>(mut self, pattern: &str, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -229,7 +229,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// of `async/await` syntax in the callback.
     pub fn delete_async<T>(mut self, pattern: &str, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -247,7 +247,7 @@ impl<'a, D: 'static> Router<'a, D> {
         func: fn(Request, RouteContext<D>) -> T,
     ) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -261,7 +261,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// syntax in the callback.
     pub fn on_async<T>(mut self, pattern: &str, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.add_handler(
             pattern,
@@ -276,7 +276,7 @@ impl<'a, D: 'static> Router<'a, D> {
     /// the callback.
     pub fn not_found_async<T>(mut self, func: fn(Request, RouteContext<D>) -> T) -> Self
     where
-        T: Future<Output = Result<Response>> + 'static,
+        T: Future<Output = Result<Response>> + 'a,
     {
         self.not_found_handler = Some(Handler::Async(Rc::new(move |req, route| {
             Box::pin(func(req, route))
@@ -350,7 +350,7 @@ impl<'a, D: 'static> Router<'a, D> {
 
 type NodeWithHandlers<'a, D> = Node<Handler<'a, D>>;
 
-impl<'a, D: 'static> Router<'a, D> {
+impl<'a, D> Router<'a, D> {
     fn split(
         self,
     ) -> (
```

Reply:
Don't you still have `'static` on `@@ -261,7 +261,7 @@ impl<'a, D: 'static> Router<'a, D> {` etc.?

rust-analyzer is yelling if I remove / change `'static` to `'a`

Reply:
and cargo check giving the same errors:

```
...
...
error[E0310]: the parameter type `D` may not live long enough
   --> worker/src/router.rs:281:54
    |
92  |   impl<'a, D: 'a> Router<'a, D> {
    |            -- help: consider adding an explicit lifetime bound...: `D: 'static +`
...
281 |           self.not_found_handler = Some(Handler::Async(Rc::new(move |req, route| {
    |  ______________________________________________________^
282 | |             Box::pin(func(req, route))
283 | |         })));
    | |__________^ ...so that the type `[closure@worker/src/router.rs:281:62: 283:10]` will meet its required lifetime bounds

For more information about this error, try `rustc --explain E0310`.
error: could not compile `worker` due to 18 previous errors
➜  worker git:(merge/pr-34) ✗ rustc --version
rustc 1.55.0 (c8dfcfe04 2021-09-06)
```

Reply:
weird -- this is working now 👻